File: Whatsapp/OpenWA/sdk/python/openwa/resources/messages.py
"""Messages resource — sending and querying messages.

Backed by ``src/modules/message/message.controller.ts``.
NOTE: the real paths use the ``/send-`` prefix, e.g. ``/messages/send-text``.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from .._http import quote_segment
from ..types import (
    BatchStatusResponse,
    BulkMessageResponse,
    ChatHistoryMessage,
    DeleteMessageRequest,
    EditMessageRequest,
    ForwardMessageRequest,
    ListMessagesQuery,
    MessageHistoryQuery,
    MessageListResponse,
    MessageResponse,
    ReactionRecord,
    ReactMessageRequest,
    ReplyMessageRequest,
    SendBulkRequest,
    SendContactRequest,
    SendLocationRequest,
    SendAudioRequest,
    SendMediaRequest,
    SendPollRequest,
    SendTemplateRequest,
    SendTextRequest,
    SuccessResult,
)

if TYPE_CHECKING:
    from .._http import HttpExecutor


class MessagesResource:
    def __init__(self, http: "HttpExecutor") -> None:
        self._http = http

    def list(self, session_id: str, query: ListMessagesQuery | None = None) -> MessageListResponse:
        return self._http.request("GET", f"/api/sessions/{quote_segment(session_id)}/messages", query=query)

    def send_text(self, session_id: str, body: SendTextRequest) -> MessageResponse:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/send-text", body=body)

    def send_image(self, session_id: str, body: SendMediaRequest) -> MessageResponse:
        return self._send_media(session_id, "send-image", body)

    def send_video(self, session_id: str, body: SendMediaRequest) -> MessageResponse:
        return self._send_media(session_id, "send-video", body)

    def send_audio(self, session_id: str, body: SendAudioRequest) -> MessageResponse:
        return self._send_media(session_id, "send-audio", body)

    def send_document(self, session_id: str, body: SendMediaRequest) -> MessageResponse:
        return self._send_media(session_id, "send-document", body)

    def send_sticker(self, session_id: str, body: SendMediaRequest) -> MessageResponse:
        return self._send_media(session_id, "send-sticker", body)

    def _send_media(self, session_id: str, segment: str, body: SendMediaRequest) -> MessageResponse:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/{quote_segment(segment)}", body=body)

    def send_location(self, session_id: str, body: SendLocationRequest) -> MessageResponse:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/send-location", body=body)

    def send_contact(self, session_id: str, body: SendContactRequest) -> MessageResponse:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/send-contact", body=body)

    def send_template(self, session_id: str, body: SendTemplateRequest) -> MessageResponse:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/send-template", body=body)

    def send_poll(self, session_id: str, body: SendPollRequest) -> MessageResponse:
        """Send a native WhatsApp poll (2–12 options)."""
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/send-poll", body=body)

    def reply(self, session_id: str, body: ReplyMessageRequest) -> MessageResponse:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/reply", body=body)

    def forward(self, session_id: str, body: ForwardMessageRequest) -> MessageResponse:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/forward", body=body)

    def react(self, session_id: str, body: ReactMessageRequest) -> SuccessResult:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/react", body=body)

    def delete(self, session_id: str, body: DeleteMessageRequest) -> SuccessResult:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/delete", body=body)

    def edit_message(self, session_id: str, body: EditMessageRequest) -> MessageResponse:
        """Edit the text of a message sent by this account. 404 when the message is not found."""
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/edit", body=body)

    def history(
        self, session_id: str, chat_id: str, query: MessageHistoryQuery | None = None
    ) -> list[ChatHistoryMessage]:
        return self._http.request(
            "GET", f"/api/sessions/{quote_segment(session_id)}/messages/{quote_segment(chat_id)}/history", query=query
        )

    def reactions(self, session_id: str, chat_id: str, message_id: str) -> list[ReactionRecord]:
        return self._http.request(
            "GET", f"/api/sessions/{quote_segment(session_id)}/messages/{quote_segment(chat_id)}/{quote_segment(message_id)}/reactions"
        )

    def send_bulk(self, session_id: str, body: SendBulkRequest) -> BulkMessageResponse:
        return self._http.request("POST", f"/api/sessions/{quote_segment(session_id)}/messages/send-bulk", body=body)

    def batch_status(self, session_id: str, batch_id: str) -> BatchStatusResponse:
        return self._http.request("GET", f"/api/sessions/{quote_segment(session_id)}/messages/batch/{quote_segment(batch_id)}")

    def cancel_batch(self, session_id: str, batch_id: str) -> BatchStatusResponse:
        """Cancel a running batch. Requires an OPERATOR-level key."""
        return self._http.request(
            "POST", f"/api/sessions/{quote_segment(session_id)}/messages/batch/{quote_segment(batch_id)}/cancel"
        )

File: OpenWA/sdk/javascript/src/resources/messages.ts
/**
 * Messages resource — sending and querying messages.
 *
 * Backed by `src/modules/message/message.controller.ts`.
 * NOTE: the real paths use the `/send-` prefix, e.g. `/messages/send-text`.
 * @packageDocumentation
 */

import { encodeSegment } from '../http.js';
import type { BinaryResponse } from '../http.js';
import type { OpenWAClient } from '../client.js';
import type {
  BatchStatusResponse,
  BulkMessageResponse,
  ChatHistoryMessage,
  DeleteMessageRequest,
  EditMessageRequest,
  ForwardMessageRequest,
  ListMessagesQuery,
  MessageHistoryQuery,
  MessageListResponse,
  MessageResponse,
  PinMessageRequest,
  ReactionRecord,
  ReactMessageRequest,
  ReplyMessageRequest,
  SendBulkRequest,
  SendContactRequest,
  SendLocationRequest,
  SendAudioRequest,
  SendMediaRequest,
  StarMessageRequest,
  UnpinMessageRequest,
  VotePollRequest,
  SendPollRequest,
  SendTemplateRequest,
  SendTextRequest,
  SuccessResult,
} from '../types.js';

export class MessagesResource {
  constructor(private readonly client: OpenWAClient) {}

  /** List messages, optionally filtered by chat or sender. Returns a `{ messages, total }` page. */
  list(sessionId: string, query?: ListMessagesQuery): Promise<MessageListResponse> {
    return this.client.request<MessageListResponse>({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages`,
      query,
    });
  }

  /** Send a text message. */
  sendText(sessionId: string, body: SendTextRequest): Promise<MessageResponse> {
    return this.client.request<MessageResponse>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/send-text`,
      body,
    });
  }

  /** Send an image (url or base64). */
  sendImage(sessionId: string, body: SendMediaRequest): Promise<MessageResponse> {
    return this.client.sendMedia(sessionId, 'send-image', body);
  }

  /** Send a video (url or base64). */
  sendVideo(sessionId: string, body: SendMediaRequest): Promise<MessageResponse> {
    return this.client.sendMedia(sessionId, 'send-video', body);
  }

  /** Send an audio file (url or base64). */
  sendAudio(sessionId: string, body: SendAudioRequest): Promise<MessageResponse> {
    return this.client.sendMedia(sessionId, 'send-audio', body);
  }

  /** Send a document (url or base64; `filename` required). */
  sendDocument(sessionId: string, body: SendMediaRequest): Promise<MessageResponse> {
    return this.client.sendMedia(sessionId, 'send-document', body);
  }

  /** Send a sticker (url or base64). */
  sendSticker(sessionId: string, body: SendMediaRequest): Promise<MessageResponse> {
    return this.client.sendMedia(sessionId, 'send-sticker', body);
  }

  /** Send a location. */
  sendLocation(sessionId: string, body: SendLocationRequest): Promise<MessageResponse> {
    return this.client.request<MessageResponse>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/send-location`,
      body,
    });
  }

  /** Send a contact card. */
  sendContact(sessionId: string, body: SendContactRequest): Promise<MessageResponse> {
    return this.client.request<MessageResponse>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/send-contact`,
      body,
    });
  }

  /** Render and send a stored message template. */
  sendTemplate(sessionId: string, body: SendTemplateRequest): Promise<MessageResponse> {
    return this.client.request<MessageResponse>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/send-template`,
      body,
    });
  }

  /** Send a native WhatsApp poll (2–12 options). */
  sendPoll(sessionId: string, body: SendPollRequest): Promise<MessageResponse> {
    return this.client.request<MessageResponse>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/send-poll`,
      body,
    });
  }

  /** Reply to a specific message. */
  reply(sessionId: string, body: ReplyMessageRequest): Promise<MessageResponse> {
    return this.client.request<MessageResponse>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/reply`,
      body,
    });
  }

  /** Forward a message to another chat. */
  forward(sessionId: string, body: ForwardMessageRequest): Promise<MessageResponse> {
    return this.client.request<MessageResponse>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/forward`,
      body,
    });
  }

  /** React to a message (empty `emoji` removes the reaction). */
  react(sessionId: string, body: ReactMessageRequest): Promise<SuccessResult> {
    return this.client.request<SuccessResult>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/react`,
      body,
    });
  }

  /** Delete a message. */
  delete(sessionId: string, body: DeleteMessageRequest): Promise<SuccessResult> {
    return this.client.request<SuccessResult>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/delete`,
      body,
    });
  }

  /** Edit the text of an own message (404 if the message is not found). */
  editMessage(sessionId: string, body: EditMessageRequest): Promise<MessageResponse> {
    return this.client.request<MessageResponse>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/edit`,
      body,
    });
  }

  /** Get the message history for a chat (read live from WhatsApp). */
  history(sessionId: string, chatId: string, query?: MessageHistoryQuery): Promise<ChatHistoryMessage[]> {
    return this.client.request<ChatHistoryMessage[]>({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/${encodeSegment(chatId)}/history`,
      query,
    });
  }

  /** Get reactions for a specific message. */
  reactions(sessionId: string, chatId: string, messageId: string): Promise<ReactionRecord[]> {
    return this.client.request<ReactionRecord[]>({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/${encodeSegment(chatId)}/${encodeSegment(messageId)}/reactions`,
    });
  }

  /**
   * Pin a message in its chat. `durationSeconds` must be 86400 (24h), 604800 (7d) or 2592000
   * (30d); it defaults to 24h server-side. In a group only admins may pin.
   */
  pin(sessionId: string, body: PinMessageRequest): Promise<SuccessResult> {
    return this.client.request<SuccessResult>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/pin`,
      body,
    });
  }

  /**
   * Cast a vote on a poll. Not supported on the Baileys engine (501).
   * `options` are the option texts, not ids.
   */
  votePoll(sessionId: string, body: VotePollRequest): Promise<SuccessResult> {
    return this.client.request<SuccessResult>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/vote-poll`,
      body,
    });
  }

  /**
   * Star or unstar a message. Best-effort on whatsapp-web.js: it silently ignores a message it
   * will not star, so a resolved call is not proof the star is set.
   */
  star(sessionId: string, body: StarMessageRequest): Promise<SuccessResult> {
    return this.client.request<SuccessResult>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/star`,
      body,
    });
  }

  /** Remove a message's pin. */
  unpin(sessionId: string, body: UnpinMessageRequest): Promise<SuccessResult> {
    return this.client.request<SuccessResult>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/unpin`,
      body,
    });
  }

  /**
   * Fetch a message's archived media bytes. Requires chat-media archiving to have been enabled on
   * the gateway when the message arrived; 404 otherwise.
   */
  media(sessionId: string, chatId: string, messageId: string): Promise<BinaryResponse> {
    return this.client.requestBytes({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/${encodeSegment(chatId)}/${encodeSegment(messageId)}/media`,
    });
  }

  /**
   * Send a batch of messages asynchronously. Returns HTTP 202 with a batch id;
   * poll the status via {@link batchStatus}.
   */
  sendBulk(sessionId: string, body: SendBulkRequest): Promise<BulkMessageResponse> {
    return this.client.request<BulkMessageResponse>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/send-bulk`,
      body,
    });
  }

  /** Poll the status/progress of a bulk send batch. */
  batchStatus(sessionId: string, batchId: string): Promise<BatchStatusResponse> {
    return this.client.request<BatchStatusResponse>({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/batch/${encodeSegment(batchId)}`,
    });
  }

  /** Cancel a running batch. Requires an OPERATOR-level key. */
  cancelBatch(sessionId: string, batchId: string): Promise<BatchStatusResponse> {
    return this.client.request<BatchStatusResponse>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/batch/${encodeSegment(batchId)}/cancel`,
    });
  }
}
